fish_completions.go

```diff
@@ -137,49 +126,65 @@ function __%[1]s_prepare_completions
 
     __%[1]s_debug "nospace: $nospace, nofiles: $nofiles"
 
-    # Important not to quote the variable for count to work
-    set numComps (count $__%[1]s_comp_results)
-    __%[1]s_debug "numComps: $numComps"
-
-    if test $numComps -eq 1; and test $nospace -ne 0
-        # To support the "nospace" directive we trick the shell
-        # by outputting an extra, longer completion.
-        __%[1]s_debug "Adding second completion to perform nospace directive"
-        set --append __%[1]s_comp_results $__%[1]s_comp_results[1].
+    # If we want to prevent a space, or if file completion is NOT disabled,
+    # we need to count the number of valid completions.
+    # To do so, we will filter on prefix as the completions we have received
+    # may not already be filtered so as to allow fish to match on different
+    # criteria than the prefix.
+    if test $nospace -ne 0; or test $nofiles -eq 0
+        set prefix (commandline -t)
+        __%[1]s_debug "prefix: $prefix"
+
+        set completions
+        for comp in $__%[1]s_comp_results
+            if test (string match -e -r -- "^$prefix" "$comp")
+                set -a completions $comp
+            end
+        end
+        set --global __%[1]s_comp_results $completions
+        __%[1]s_debug "Filtered completions are: $__%[1]s_comp_results"
+
+        # Important not to quote the variable for count to work
+        set numComps (count $__%[1]s_comp_results)
+        __%[1]s_debug "numComps: $numComps"
+
+        if test $numComps -eq 1; and test $nospace -ne 0
+            # To support the "nospace" directive we trick the shell
+            # by outputting an extra, longer completion.
+            # We must first split on \t to get rid of the descriptions because
+            # the extra character we add to the fake second completion must be
+            # before the description.  We don't need descriptions anyway since
```

Comment:
This is clever, and works reliably as far as I can tell.
However, the upcoming fish 3.2.0 will always show the pager on ambiguous completions (https://github.com/fish-shell/fish-shell/issues/6924). This would cause some confusion.

So I'm not sure if this an overall improvement. Who are the users of this option?
Fish already gets this right a lot of times by hardcoding the character set `@=/:.,` - if a completion ends in one of these characters, then no space will be inserted.

Reply:
> This is clever, and works reliably as far as I can tell.
> However, the upcoming fish 3.2.0 will always show the pager on ambiguous completions ([fish-shell/fish-shell#6924](https://github.com/fish-shell/fish-shell/issues/6924)). This would cause some confusion.

Ouch, that will indeed cause confusion for the "nospace" case.

> So I'm not sure if this an overall improvement. Who are the users of this option?

For example, Helm uses this option when completing a chart's full name which is comprised of `<repository>/<chartname>`; the completion first provides all repository names in the form `<repository>/`, and once the slash is present on the command-line, the completion will add the possible chart names, but there should be no space between the two.  This is much more efficient for helm than having to provide all the full `<repository>/<chartname>` from the very start.

`kubectl cp <pod>:<file>` could also use this option to avoid adding a space after the `:`.

> Fish already gets this right a lot of times by hardcoding the character set `@=/:.,` - if a completion ends in one of these characters, then no space will be inserted.

Cool!  I was not aware of this.  This would address both cases I mentioned.  I tested it and it works as you describe.
But I don't know if there are other cases that wouldn't end with `@=/:.,`.  At a minimum, we could turn off this hack for if the completion ended with `@=/:.,`.  I'll have to think what to do for the other (possible) cases for fish 3.2.  Let me know if you have any suggestions.

Reply:
One could sidestep the completion engine and directly insert the unique completion if nospace is set

```fish
$ function sneaky-complete
      commandline --replace -t nospace
  end
$ complete -e sneaky-complete; complete -c sneaky-complete -a '(sneaky-complete)' -f
```

According to my tests this works (although it rings the terminal bell, because fish thinks there are no completions). However, I'd advise against it. You are not meant to modifiy the commandline *during* the computation of a completion. That can break other completions, though I didn't find an example here.

fish has not added an option to control the add-a-space behavior yet because AFAIK no real use case has been brought up that doesn't work with the hardcoded list.

I fear that the cost of adding a hack to support this outweighs the benefits.
I think we should drop this until we see a real need. Then an option can be added to fish, avoiding the need for a hacky workaround.

Reply:
I took a look how I use the `ShellCompDirectiveNoSpace` option in podman. Almost all cases ends with one of `=:,` but not all. I found one case where I set this when it ends with a `-`. https://github.com/containers/podman/blob/479fc226044b745ecaafc2d5fa925afe6ca06de0/cmd/podman/common/completion.go#L603-L608
Even more importantly there is a case where I set this together with `ShellCompDirectiveDefault` (path completion). This case does not work with fish currently. It works for bash and zsh.

Reply:
Thanks for investigating. I guess we should just [add `-`](https://i.redd.it/vtttn02gchsz.jpg) to the set of separators in fish.

Regarding the nospace+path, is this what you mean:

```go
// cobra-completion-testing/testprog/testprog.go
var noSpaceCmdPrefix = &cobra.Command{
	Use:   "nospace",
	Short: "Directive: no space",
	ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
		return nil, cobra.ShellCompDirectiveNoSpace
	},
	Run: func(cmd *cobra.Command, args []string) {},
}
```

```
testprog/bin/testprog prefix nospace <TAB>
```

this gives me files in both fish and bash (I've merged this branch and the bash v2 one).
If I return some completions instead of `nil`, then I only get those completions in both shells (as expected).

You probably mean something else?

Reply:
Bash does not add a space after the completed path. Fish does.

Reply:
Honestly I don't think adding `-` make much sense. This might solve my specific problem but doesn't solve the problem here.
`cobra` allows me as developer to set `nospace` wherever I want, not only on those characters. I don't see a reason why fish should limit my options as a completion script developer.

Reply:
> Even more importantly there is a case where I set this together with ShellCompDirectiveDefault (path completion). 
> Bash does not add a space after the completed path. Fish does.

@Luap99 I never planned for this case.  And now that you mention it, I realize it also does not work for `zsh`.  It is an easy fix though.  I will add it to #1213.  And I see that you made it work for powershell 🎉 .  I don't believe there is anything we can do for `fish` though.  

Out of curiosity, what kind of situation would not want a space after file completion?

Reply:
Turns out preventing a space after file completion for `zsh` also prevents the trailing `/` from being added after a directory.  At least in the way I've found to prevent a space.  I think a missing `/` is a bigger problem than the extra space, so I'll leave it alone.

Reply:
To get back to the 'nospace' directive for fish, the latest version of this PR keeps the "hack" that adds an extra fake completion but only when the completion does not end with `@=/:.,`.  I think this is the best we can do at this time.

Reply:
> Out of curiosity, what kind of situation would not want a space after file completion?

Simple answer `podman/docker run --volume`. This flag accepts a volume name or a path followed by a colon and again a path. This is my current autocompletion function:
```
func AutocompleteVolumeFlag(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	volumes, _ := getVolumes(cmd, toComplete)
	directive := cobra.ShellCompDirectiveNoSpace | cobra.ShellCompDirectiveDefault
	if strings.Contains(toComplete, ":") {
		// add space after second path
		directive = cobra.ShellCompDirectiveDefault
	}
	return volumes, directive
}
```
After the first path I do not want the space added. The user has to type a colon here. After the colon you get only path completion and then I a space should be added.

> Turns out preventing a space after file completion for zsh also prevents the trailing / from being added after a directory. At least in the way I've found to prevent a space. I think a missing / is a bigger problem than the extra space, so I'll leave it alone.

I am totally fine if it does not work. It is just a small corner case.